File: platforms/apple/Sources/BotaAppleSDK/Host/FileRecordingSinkHost.swift
import BotaDeviceSDKC
import Foundation

actor FileRecordingSinkHost: RecordingSinkHost {
    typealias StreamingUpload = @Sendable (StreamingUploadDestination, Data) async throws -> Void

    private let rootDirectory: URL
    private let fileManager: FileManager
    private let streamingUpload: StreamingUpload
    private var preparedSinks: Set<String> = []
    private var streamingSinks: [String: StreamingSinkSession] = [:]

    init(
        rootDirectory: URL,
        fileManager: FileManager = .default,
        streamingUpload: @escaping StreamingUpload = FileRecordingSinkHost.upload
    ) {
        self.rootDirectory = rootDirectory
        self.fileManager = fileManager
        self.streamingUpload = streamingUpload
    }

    func registerStreaming(
        sinkID: String,
        chunkSizeBytes: Int,
        flushIntervalMilliseconds: UInt64,
        destinationProvider: @escaping StreamingChunkDestinationProvider,
        finalize: @escaping StreamingFinalizeHandler
    ) throws {
        guard UUID(uuidString: sinkID) != nil else { throw NativeHostError.invalidOpaqueID(sinkID) }
        guard chunkSizeBytes > 0 else { throw NativeHostError.missingResource("streaming chunk size") }
        streamingSinks[sinkID] = StreamingSinkSession(
            chunkSizeBytes: chunkSizeBytes,
            flushIntervalMilliseconds: flushIntervalMilliseconds,
            destinationProvider: destinationProvider,
            finalize: finalize,
            upload: streamingUpload
        )
    }

    func unregisterStreaming(sinkID: String) async {
        if let session = streamingSinks.removeValue(forKey: sinkID) {
            await session.discard()
        }
    }

    func execute(_ effect: CoreEffect) async -> AsyncThrowingStream<CoreHostEventPayload, Error> {
        if effect.isStreamingEffect {
            do {
                let sinkID = try requiredText(effect, UInt32(BOTA_DEVICE_SDK_V1_FIELD_SINK_ID))
                guard let session = streamingSinks[sinkID] else { throw NativeHostError.missingResource(sinkID) }
                let payload = try await session.execute(effect)
                return AsyncThrowingStream { continuation in
                    if let payload { continuation.yield(payload) }
                    continuation.finish()
                }
            } catch {
                return AsyncThrowingStream { $0.finish(throwing: error) }
            }
        }
        return AsyncThrowingStream<CoreHostEventPayload, Error> { continuation in
            do {
                let sinkID = try requiredText(effect, UInt32(BOTA_DEVICE_SDK_V1_FIELD_SINK_ID))
                let url = try sinkURL(sinkID)
                switch effect {
                case .recordingSinkTruncate:
                    let completed = try requiredUnsigned(effect, UInt32(BOTA_DEVICE_SDK_V1_FIELD_COMPLETED_UNITS))
                    try truncate(url, to: completed)
                    preparedSinks.insert(sinkID)
                    continuation.yield(.init(kind: UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_RECORDING_SINK_TRUNCATED)))
                case .recordingSinkAppend:
                    guard preparedSinks.contains(sinkID) else { throw NativeHostError.missingResource(sinkID) }
                    let payload = try requiredBytes(effect, UInt32(BOTA_DEVICE_SDK_V1_FIELD_PAYLOAD))
                    let durableUnits = try append(payload, to: url)
                    continuation.yield(.init(
                        kind: UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_RECORDING_SINK_APPEND_COMPLETED),
                        fields: [.unsigned(
                            id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_DURABLE_UNITS),
                            value: durableUnits
                        )]
                    ))
                case .recordingSinkFinalize:
                    guard preparedSinks.contains(sinkID) else { throw NativeHostError.missingResource(sinkID) }
                    let durableUnits = try fileSize(url)
                    if let expected = optionalUnsigned(effect, UInt32(BOTA_DEVICE_SDK_V1_FIELD_EXPECTED_CRC32)),
                       UInt64(try crc32(url)) != expected {
                        continuation.yield(.init(
                            kind: UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_RECORDING_SINK_INTEGRITY_FAILED)
                        ))
                    } else {
                        continuation.yield(.init(
                            kind: UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_RECORDING_SINK_FINALIZED),
                            fields: [.unsigned(
                                id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_DURABLE_UNITS),
                                value: durableUnits
                            )]
                        ))
                    }
                case .recordingSinkDiscard:
                    try removeIfPresent(url)
                    preparedSinks.remove(sinkID)
                default:
                    throw NativeHostError.invalidEffect(effect.kind)
                }
                continuation.finish()
            } catch {
                continuation.finish(throwing: error)
            }
        }
    }

    func fileURL(for sinkID: String) throws -> URL { try sinkURL(sinkID) }

    private func sinkURL(_ sinkID: String) throws -> URL {
        guard UUID(uuidString: sinkID) != nil else { throw NativeHostError.invalidOpaqueID(sinkID) }
        return rootDirectory.appendingPathComponent(sinkID).appendingPathExtension("recording")
    }

    private func truncate(_ url: URL, to completed: UInt64) throws {
        try fileManager.createDirectory(at: rootDirectory, withIntermediateDirectories: true)
        if !fileManager.fileExists(atPath: url.path) { fileManager.createFile(atPath: url.path, contents: nil) }
        let handle = try FileHandle(forWritingTo: url)
        defer { try? handle.close() }
        try handle.truncate(atOffset: completed)
        try handle.synchronize()
    }

    private func append(_ data: Data, to url: URL) throws -> UInt64 {
        let handle = try FileHandle(forWritingTo: url)
        defer { try? handle.close() }
        try handle.seekToEnd()
        try handle.write(contentsOf: data)
        try handle.synchronize()
        return try handle.offset()
    }

    private func fileSize(_ url: URL) throws -> UInt64 {
        let values = try url.resourceValues(forKeys: [.fileSizeKey])
        return UInt64(values.fileSize ?? 0)
    }

    private func crc32(_ url: URL) throws -> UInt32 {
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }
        var checksum: UInt32 = 0xFFFF_FFFF
        while let chunk = try handle.read(upToCount: 64 * 1024), !chunk.isEmpty {
            for byte in chunk {
                checksum ^= UInt32(byte)
                for _ in 0..<8 {
                    checksum = (checksum >> 1) ^ ((checksum & 1) == 1 ? 0xEDB8_8320 : 0)
                }
            }
        }
        return checksum ^ 0xFFFF_FFFF
    }

    private func removeIfPresent(_ url: URL) throws {
        if fileManager.fileExists(atPath: url.path) { try fileManager.removeItem(at: url) }
    }

    private static func upload(destination: StreamingUploadDestination, body: Data) async throws {
        var request = URLRequest(url: destination.url)
        request.httpMethod = destination.method.rawValue
        request.setValue(destination.contentType, forHTTPHeaderField: "Content-Type")
        if let bearerToken = destination.bearerToken {
            request.setValue("Bearer \(bearerToken)", forHTTPHeaderField: "Authorization")
        }
        let (_, response) = try await URLSession.shared.upload(for: request, from: body)
        guard let http = response as? HTTPURLResponse,
              (200 ... 299).contains(http.statusCode)
        else { throw NativeHostError.httpStatus((response as? HTTPURLResponse)?.statusCode ?? 0) }
    }
}

private actor StreamingSinkSession {
    private let chunkSizeBytes: Int
    private let flushIntervalMilliseconds: UInt64
    private let destinationProvider: StreamingChunkDestinationProvider
    private let finalizeHandler: StreamingFinalizeHandler
    private let uploadHandler: FileRecordingSinkHost.StreamingUpload
    private let startedAtNanoseconds = DispatchTime.now().uptimeNanoseconds
    private var plaintextBuffer = Data()
    private var encryptedHeader: Data?
    private var encrypted: Bool?
    private var completedUnits: UInt64 = 0
    private var uploadedChunks: UInt32 = 0
    private var nextPlaintextSequence: UInt32 = 1
    private var flushTask: Task<Void, Never>?
    private var backgroundFailure: Error?
    private var discarded = false

    init(
        chunkSizeBytes: Int,
        flushIntervalMilliseconds: UInt64,
        destinationProvider: @escaping StreamingChunkDestinationProvider,
        finalize: @escaping StreamingFinalizeHandler,
        upload: @escaping FileRecordingSinkHost.StreamingUpload
    ) {
        self.chunkSizeBytes = chunkSizeBytes
        self.flushIntervalMilliseconds = flushIntervalMilliseconds
        self.destinationProvider = destinationProvider
        self.finalizeHandler = finalize
        self.uploadHandler = upload
    }

    func execute(_ effect: CoreEffect) async throws -> CoreHostEventPayload? {
        guard !discarded else { throw NativeHostError.missingResource("discarded streaming sink") }
        if let backgroundFailure { throw backgroundFailure }
        switch effect {
        case .streamingSinkAppendPlaintext:
            guard encrypted != true else { throw NativeHostError.invalidEffect(effect.kind) }
            encrypted = false
            let payload = try requiredBytes(effect, UInt32(BOTA_DEVICE_SDK_V1_FIELD_PAYLOAD))
            completedUnits += UInt64(payload.count)
            plaintextBuffer.append(payload)
            try await flushFullPlaintextChunks()
            schedulePartialFlush()
            return accepted()
        case .streamingSinkBeginEncrypted:
            guard encrypted != false else { throw NativeHostError.invalidEffect(effect.kind) }
            let key = try requiredBytes(effect, UInt32(BOTA_DEVICE_SDK_V1_FIELD_EPHEMERAL_PUBLIC_KEY))
            let salt = try requiredBytes(effect, UInt32(BOTA_DEVICE_SDK_V1_FIELD_SALT))
            guard key.count == 32, salt.count == 4 else { throw NativeHostError.invalidEffect(effect.kind) }
            encrypted = true
            encryptedHeader = key + salt
            return accepted()
        case .streamingSinkAppendEncrypted:
            guard encrypted == true, let encryptedHeader else { throw NativeHostError.invalidEffect(effect.kind) }
            let sequence = try requiredUnsigned(effect, UInt32(BOTA_DEVICE_SDK_V1_FIELD_SEQUENCE))
            guard sequence <= UInt64(UInt32.max) else { throw NativeHostError.invalidEffect(effect.kind) }
            let payload = try requiredBytes(effect, UInt32(BOTA_DEVICE_SDK_V1_FIELD_PAYLOAD))
            guard payload.count >= 16 else { throw NativeHostError.invalidEffect(effect.kind) }
            let request = StreamingChunkRequest(sequence: UInt32(sequence), isEncrypted: true)
            let body = sequence == 0 ? encryptedHeader + payload : payload
            do {
                try await uploadWithRetry(request: request, body: body)
                uploadedChunks += 1
            } catch where sequence > 0 {
                // The backend finalizer preserves the firmware sequence as an explicit audio gap.
            }
            completedUnits += UInt64(payload.count - 16)
            return accepted()
        case .streamingSinkFinalize:
            flushTask?.cancel()
            flushTask = nil
            if let backgroundFailure { throw backgroundFailure }
            let isEncrypted = try requiredBool(effect, UInt32(BOTA_DEVICE_SDK_V1_FIELD_ENCRYPTED))
            let expectedChunks = try requiredUnsigned(effect, 125)
            let totalUnits = try requiredUnsigned(effect, UInt32(BOTA_DEVICE_SDK_V1_FIELD_TOTAL_UNITS))
            guard encrypted == isEncrypted, totalUnits == completedUnits,
                  expectedChunks <= UInt64(UInt32.max)
            else { throw NativeHostError.invalidEffect(effect.kind) }
            if !isEncrypted { try await flushPartialPlaintext() }
            let finalizedChunks = isEncrypted ? UInt32(expectedChunks) : uploadedChunks
            let milliseconds = (
                DispatchTime.now().uptimeNanoseconds &- startedAtNanoseconds
            ) / 1_000_000
            try await finalizeHandler(.init(
                totalChunks: finalizedChunks,
                durationMilliseconds: milliseconds,
                fileSizeBytes: completedUnits,
                isEncrypted: isEncrypted
            ))
            return .init(kind: 0x0229, fields: [
                .unsigned(id: 126, value: UInt64(uploadedChunks)),
                .unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_TOTAL_UNITS), value: completedUnits),
            ])
        case .streamingSinkDiscard:
            discard()
            return nil
        default:
            throw NativeHostError.invalidEffect(effect.kind)
        }
    }

    func discard() {
        discarded = true
        flushTask?.cancel()
        flushTask = nil
        plaintextBuffer.removeAll(keepingCapacity: false)
        encryptedHeader = nil
    }

    private func accepted() -> CoreHostEventPayload {
        .init(kind: 0x0228, fields: [
            .unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_COMPLETED_UNITS), value: completedUnits),
        ])
    }

    private func flushFullPlaintextChunks() async throws {
        while plaintextBuffer.count >= chunkSizeBytes {
            let body = Data(plaintextBuffer.prefix(chunkSizeBytes))
            plaintextBuffer.removeFirst(chunkSizeBytes)
            try await uploadPlaintext(body)
        }
    }

    private func flushPartialPlaintext() async throws {
        guard !plaintextBuffer.isEmpty else { return }
        let body = plaintextBuffer
        plaintextBuffer.removeAll(keepingCapacity: true)
        try await uploadPlaintext(body)
    }

    private func uploadPlaintext(_ body: Data) async throws {
        let sequence = nextPlaintextSequence
        nextPlaintextSequence += 1
        try await uploadWithRetry(
            request: .init(sequence: sequence, isEncrypted: false),
            body: body
        )
        uploadedChunks += 1
    }

    private func uploadWithRetry(request: StreamingChunkRequest, body: Data) async throws {
        var lastError: Error?
        for attempt in 0..<3 {
            do {
                let destination = try await destinationProvider(request)
                try await uploadHandler(destination, body)
                return
            } catch {
                lastError = error
                if attempt < 2 {
                    try await Task.sleep(nanoseconds: UInt64(50 * (attempt + 1)) * 1_000_000)
                }
            }
        }
        throw lastError ?? NativeHostError.missingResource("streaming upload")
    }

    private func schedulePartialFlush() {
        flushTask?.cancel()
        guard flushIntervalMilliseconds > 0, !plaintextBuffer.isEmpty else { return }
        let delay = flushIntervalMilliseconds
        flushTask = Task { [weak self] in
            do {
                try await Task.sleep(nanoseconds: delay * 1_000_000)
                try await self?.flushPartialPlaintext()
            } catch is CancellationError {
            } catch {
                await self?.recordBackgroundFailure(error)
            }
        }
    }

    private func recordBackgroundFailure(_ error: Error) {
        backgroundFailure = error
    }

    private func requiredBool(_ effect: CoreEffect, _ id: UInt32) throws -> Bool {
        for field in effect.packet.fields {
            if case let .bool(fieldID, value) = field, fieldID == id { return value }
        }
        throw NativeHostError.missingField(id)
    }
}

private extension CoreEffect {
    var isStreamingEffect: Bool {
        switch self {
        case .streamingSinkAppendPlaintext, .streamingSinkBeginEncrypted,
             .streamingSinkAppendEncrypted, .streamingSinkFinalize,
             .streamingSinkDiscard:
            true
        default:
            false
        }
    }
}

File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/internal/host/FileRecordingSinkHost.kt
package dev.bota.sdk.internal.host

import android.os.ParcelFileDescriptor
import dev.bota.sdk.internal.core.CoreEffect
import dev.bota.sdk.internal.core.CoreEffectKind
import dev.bota.sdk.internal.core.CoreField
import dev.bota.sdk.internal.core.HostEventKind
import dev.bota.sdk.model.StreamingChunkDestinationProvider
import dev.bota.sdk.model.StreamingChunkRequest
import dev.bota.sdk.model.StreamingFinalizeHandler
import dev.bota.sdk.model.StreamingFinalizeMetadata
import dev.bota.sdk.model.StreamingUploadDestination
import java.nio.ByteBuffer
import java.nio.file.Path
import java.util.zip.CRC32
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

internal class FileRecordingSinkHost(
    private val registry: FileResourceRegistry = FileResourceRegistry(),
    private val networkClient: OkHttpClient = OkHttpClient(),
    private val streamingUpload: suspend (StreamingUploadDestination, ByteArray) -> Unit = { destination, body ->
        upload(networkClient, destination, body)
    },
) : RecordingSinkHost, AutoCloseable {
    private val mutex = Mutex()
    private val prepared = mutableSetOf<String>()
    private val streaming = mutableMapOf<String, StreamingSinkSession>()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    fun registerPath(sinkId: String, path: Path) = registry.registerPath(sinkId, path)

    fun registerDescriptor(sinkId: String, descriptor: ParcelFileDescriptor) =
        registry.registerDescriptor(sinkId, descriptor)

    fun unregister(sinkId: String) {
        prepared.remove(sinkId)
        registry.remove(sinkId)
    }

    suspend fun registerStreaming(
        sinkId: String,
        chunkSizeBytes: Int,
        flushIntervalMilliseconds: ULong,
        destinationProvider: StreamingChunkDestinationProvider,
        finalize: StreamingFinalizeHandler,
    ) {
        validOpaqueId(sinkId)
        require(chunkSizeBytes > 0) { "streaming chunk size must be positive" }
        mutex.withLock {
            streaming.remove(sinkId)?.discard()
            streaming[sinkId] = StreamingSinkSession(
                chunkSizeBytes,
                flushIntervalMilliseconds,
                destinationProvider,
                finalize,
            )
        }
    }

    suspend fun unregisterStreaming(sinkId: String) {
        mutex.withLock { streaming.remove(sinkId)?.discard() }
    }

    override fun execute(effect: CoreEffect): Flow<CoreHostEventPayload> = flow {
        mutex.withLock {
            val sinkId = validOpaqueId(requiredText(effect, HostFieldId.SinkId))
            when (effect.kind) {
                CoreEffectKind.RecordingSinkTruncate -> {
                    val resource = registry.resource(sinkId)
                    val completed = requiredUnsigned(effect, HostFieldId.CompletedUnits).toLongChecked()
                    resource.openWrite().use { channel ->
                        channel.truncate(completed)
                        channel.force(true)
                    }
                    prepared += sinkId
                    emit(CoreHostEventPayload(HostEventKind.RecordingSinkTruncated))
                }
                CoreEffectKind.RecordingSinkAppend -> {
                    val resource = registry.resource(sinkId)
                    requirePrepared(sinkId)
                    val payload = requiredBytes(effect, HostFieldId.Payload)
                    val durable = resource.openWrite().use { channel ->
                        channel.position(channel.size())
                        val buffer = ByteBuffer.wrap(payload)
                        while (buffer.hasRemaining()) channel.write(buffer)
                        channel.force(true)
                        channel.position().toULong()
                    }
                    emit(
                        CoreHostEventPayload(
                            HostEventKind.RecordingSinkAppendCompleted,
                            listOf(CoreField.Unsigned(HostFieldId.DurableUnits, durable)),
                        ),
                    )
                }
                CoreEffectKind.RecordingSinkFinalize -> {
                    val resource = registry.resource(sinkId)
                    requirePrepared(sinkId)
                    val size = resource.openRead().use(FileChannelSize)
                    val expected = effect.packet.unsigneds(HostFieldId.ExpectedCrc32).firstOrNull()
                    val kind = if (expected != null && crc32(resource).toULong() != expected) {
                        HostEventKind.RecordingSinkIntegrityFailed
                    } else {
                        HostEventKind.RecordingSinkFinalized
                    }
                    val fields = if (kind == HostEventKind.RecordingSinkFinalized) {
                        listOf(CoreField.Unsigned(HostFieldId.DurableUnits, size))
                    } else {
                        emptyList()
                    }
                    emit(CoreHostEventPayload(kind, fields))
                }
                CoreEffectKind.RecordingSinkDiscard -> {
                    prepared -= sinkId
                    registry.remove(sinkId, discard = true)
                }
                CoreEffectKind.StreamingSinkAppendPlaintext -> {
                    val session = requiredStreaming(sinkId)
                    val completed = session.appendPlaintext(requiredBytes(effect, HostFieldId.Payload))
                    emit(streamingAccepted(completed))
                }
                CoreEffectKind.StreamingSinkBeginEncrypted -> {
                    val session = requiredStreaming(sinkId)
                    session.beginEncrypted(
                        requiredBytes(effect, HostFieldId.EphemeralPublicKey),
                        requiredBytes(effect, HostFieldId.Salt),
                    )
                    emit(streamingAccepted(session.completedUnits))
                }
                CoreEffectKind.StreamingSinkAppendEncrypted -> {
                    val session = requiredStreaming(sinkId)
                    val sequence = requiredUnsigned(effect, HostFieldId.Sequence).toUIntChecked()
                    val completed = session.appendEncrypted(sequence, requiredBytes(effect, HostFieldId.Payload))
                    emit(streamingAccepted(completed))
                }
                CoreEffectKind.StreamingSinkFinalize -> {
                    val session = requiredStreaming(sinkId)
                    val result = session.finalize(
                        encrypted = requiredBoolean(effect, HostFieldId.Encrypted),
                        expectedChunks = requiredUnsigned(effect, HostFieldId.ExpectedChunks).toUIntChecked(),
                        totalUnits = requiredUnsigned(effect, HostFieldId.TotalUnits),
                    )
                    emit(
                        CoreHostEventPayload(
                            HostEventKind.StreamingSinkFinalized,
                            listOf(
                                CoreField.Unsigned(HostFieldId.UploadedChunks, result.first.toULong()),
                                CoreField.Unsigned(HostFieldId.TotalUnits, result.second),
                            ),
                        ),
                    )
                }
                CoreEffectKind.StreamingSinkDiscard -> {
                    streaming.remove(sinkId)?.discard()
                }
                else -> throw NativeHostException(422, "non-recording effect reached recording sink")
            }
        }
    }.flowOn(Dispatchers.IO)

    override fun close() {
        scope.cancel()
        streaming.values.forEach(StreamingSinkSession::discard)
        streaming.clear()
        prepared.clear()
        registry.close()
    }

    private fun requiredStreaming(sinkId: String): StreamingSinkSession =
        streaming[sinkId] ?: throw NativeHostException(404, "streaming sink is not registered: $sinkId")

    private fun streamingAccepted(completed: ULong): CoreHostEventPayload = CoreHostEventPayload(
        HostEventKind.StreamingSinkAccepted,
        listOf(CoreField.Unsigned(HostFieldId.CompletedUnits, completed)),
    )

    private fun requirePrepared(sinkId: String) {
        if (sinkId !in prepared) throw NativeHostException(409, "recording sink was not prepared")
    }

    private fun crc32(resource: FileResource): Long {
        val checksum = CRC32()
        resource.openRead().use { channel ->
            val buffer = ByteBuffer.allocate(64 * 1024)
            while (channel.read(buffer) >= 0) {
                if (buffer.position() == 0) continue
                checksum.update(buffer.array(), 0, buffer.position())
                buffer.clear()
            }
        }
        return checksum.value
    }

    private companion object {
        val FileChannelSize: (java.nio.channels.FileChannel) -> ULong = { it.size().toULong() }

        suspend fun upload(
            client: OkHttpClient,
            destination: StreamingUploadDestination,
            body: ByteArray,
        ) {
            val requestBody = body.toRequestBody(destination.contentType.toMediaType())
            val request = Request.Builder()
                .url(destination.url)
                .method(destination.method.wireValue, requestBody)
                .apply {
                    destination.bearerToken?.let { header("Authorization", "Bearer $it") }
                }
                .build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw NativeHostException(response.code, "streaming upload failed with HTTP ${response.code}")
                }
            }
        }
    }

    private inner class StreamingSinkSession(
        private val chunkSizeBytes: Int,
        private val flushIntervalMilliseconds: ULong,
        private val destinationProvider: StreamingChunkDestinationProvider,
        private val finalizeHandler: StreamingFinalizeHandler,
    ) {
        private val startedAtMilliseconds = System.currentTimeMillis()
        private var plaintextBuffer = ByteArray(0)
        private var encryptedHeader: ByteArray? = null
        private var encrypted: Boolean? = null
        var completedUnits: ULong = 0u
            private set
        private var uploadedChunks: UInt = 0u
        private var nextPlaintextSequence: UInt = 1u
        private var flushJob: Job? = null
        private var backgroundFailure: Throwable? = null
        private var discarded = false

        suspend fun appendPlaintext(payload: ByteArray): ULong {
            checkActive()
            check(encrypted != true) { "streaming sink mixed plaintext and encrypted payloads" }
            encrypted = false
            plaintextBuffer += payload
            completedUnits += payload.size.toULong()
            while (plaintextBuffer.size >= chunkSizeBytes) {
                val body = plaintextBuffer.copyOfRange(0, chunkSizeBytes)
                plaintextBuffer = plaintextBuffer.copyOfRange(chunkSizeBytes, plaintextBuffer.size)
                uploadPlaintext(body)
            }
            scheduleFlush()
            return completedUnits
        }

        fun beginEncrypted(ephemeralPublicKey: ByteArray, salt: ByteArray) {
            checkActive()
            check(encrypted != false && ephemeralPublicKey.size == 32 && salt.size == 4) {
                "invalid encrypted streaming header"
            }
            encrypted = true
            encryptedHeader = ephemeralPublicKey + salt
        }

        suspend fun appendEncrypted(sequence: UInt, payload: ByteArray): ULong {
            checkActive()
            val header = checkNotNull(encryptedHeader) { "encrypted streaming header is missing" }
            check(encrypted == true && payload.size >= 16) { "invalid encrypted streaming payload" }
            val body = if (sequence == 0u) header + payload else payload
            runCatching { uploadWithRetry(StreamingChunkRequest(sequence, true), body) }
                .onSuccess { uploadedChunks += 1u }
                .onFailure { if (sequence == 0u) throw it }
            completedUnits += (payload.size - 16).toULong()
            return completedUnits
        }

        suspend fun finalize(
            encrypted: Boolean,
            expectedChunks: UInt,
            totalUnits: ULong,
        ): Pair<UInt, ULong> {
            checkActive()
            flushJob?.cancel()
            flushJob = null
            backgroundFailure?.let { throw it }
            check(this.encrypted == encrypted && completedUnits == totalUnits) {
                "streaming finalize metadata does not match accepted bytes"
            }
            if (!encrypted) flushPartial()
            val finalizedChunks = if (encrypted) expectedChunks else uploadedChunks
            finalizeHandler.finalize(
                StreamingFinalizeMetadata(
                    finalizedChunks,
                    (System.currentTimeMillis() - startedAtMilliseconds).coerceAtLeast(0).toULong(),
                    completedUnits,
                    encrypted,
                ),
            )
            return uploadedChunks to completedUnits
        }

        fun discard() {
            discarded = true
            flushJob?.cancel()
            flushJob = null
            plaintextBuffer = ByteArray(0)
            encryptedHeader = null
        }

        private suspend fun flushPartial() {
            if (plaintextBuffer.isEmpty()) return
            val body = plaintextBuffer
            plaintextBuffer = ByteArray(0)
            uploadPlaintext(body)
        }

        private suspend fun uploadPlaintext(body: ByteArray) {
            val sequence = nextPlaintextSequence++
            uploadWithRetry(StreamingChunkRequest(sequence, false), body)
            uploadedChunks += 1u
        }

        private suspend fun uploadWithRetry(request: StreamingChunkRequest, body: ByteArray) {
            var failure: Throwable? = null
            repeat(3) { attempt ->
                try {
                    streamingUpload(destinationProvider.destination(request), body)
                    return
                } catch (error: Throwable) {
                    failure = error
                    if (attempt < 2) delay(50L * (attempt + 1))
                }
            }
            throw failure ?: NativeHostException(1, "streaming upload failed")
        }

        private fun scheduleFlush() {
            flushJob?.cancel()
            if (flushIntervalMilliseconds == 0uL || plaintextBuffer.isEmpty()) return
            flushJob = scope.launch {
                delay(flushIntervalMilliseconds.coerceAtMost(Long.MAX_VALUE.toULong()).toLong())
                mutex.withLock {
                    runCatching { flushPartial() }.exceptionOrNull()?.let { backgroundFailure = it }
                }
            }
        }

        private fun checkActive() {
            check(!discarded) { "streaming sink was discarded" }
            backgroundFailure?.let { throw it }
        }
    }
}

internal fun ULong.toLongChecked(): Long {
    if (this > Long.MAX_VALUE.toULong()) throw NativeHostException(422, "file offset is too large")
    return toLong()
}

private fun ULong.toUIntChecked(): UInt {
    if (this > UInt.MAX_VALUE.toULong()) throw NativeHostException(422, "unsigned value is too large")
    return toUInt()
}

private fun requiredBoolean(effect: CoreEffect, id: Int): Boolean =
    effect.packet.booleans(id).firstOrNull()
        ?: throw NativeHostException(422, "required boolean field $id is missing")
